### ISSUE TITLE
Feature request: echo function for debugging [$50]

### ISSUE THREAD

**Issue by MichaelAtOz** [openscad/openscad#741]
Debugging recursive modules is pretty straight forward, recursive functions is a little harder, particularly if you intertwine two such functions.

With modules, echo helps.

So how about an echof() function; or just echo() if there isn't an issue with functions & modules having the same name

echof(same parameters as echo statement) returns zero.

Then you can use it in numeric expressions, or with || added a boolean expression (left side), or add as another parameter to a function call.

```
function echof() = 0; // mockup

function recur(i=0) = recur(i+1, echof("i=",i) );  // yes bad recursion

function list(j=0,i=0) =
    ( echof("logic expr, j=",j) || (i==j) )
     ? [] 
     : concat(i,
             list(j
                  + echof("numeric expr, j=",j)  // layout allows echo to be commented out
                  ,i+1, 
                  echof("extra param, i=",i)
                 ) // yes overkill
             ); // but demo of the options

echo(list(5));
```

May need to ensure the echo events are processed immediately, but that may be a good thing anyway. 

Hopefully this is possible within the app architecture.
##

**Comment by MichaelAtOz** [comment]
Covered in #1859. Closing.

**Comment by MichaelAtOz** [comment]
Reopening to correctly attribute the Bounty.

**Comment by t-paul** [comment]
Closed via #1830 and #2504.
